### PR TITLE
Custom triton batcher for corvo

### PR DESCRIPTION
**Pull request by sfc-gh-bprosnitz**
This introduces the CorvoBatcher, which prevents merging of batches and allows a fixed number of outstanding requests (otherwise rejecting with "unavailable" like the dynamic batcher).

This should be functionally similar to the PR which implemented similar constraints via the [dynamic batcher](https://github.com/neevaco/neeva/pull/106482), but uses less complex code so is simpler to evaluate from a security perspective and is not prone to changing security properties due to surprise behavior changes from an outside third party.

I tested this Corvo+Triton built with this PR and many concurrent requests.

### CODE REVIEW

File: docker/Dockerfile [hunk 1]
```diff
@@ -18,28 +18,28 @@ FROM ${BASE_IMAGE}
...
+    tmux \
+    unzip \
+    zstd \
+    zip \
+    zsh && \
     apt-get clean && \
     rm -rf /var/lib/apt/lists/*
 
-RUN pip3 install --no-cache-dir --extra-index-url https://download.pytorch.org/whl/cu118 torch==2.0.1+cu118 && \
+RUN pip3 install --no-cache-dir --extra-index-url https://download.pytorch.org/whl/cu118 torch==2.0.1 && \
```
Comment (sfc-gh-bprosnitz) [inline docker/Dockerfile:42]: When you build without this change, you get an error that 2.0.1+cu118 doesn't exist

File: docker/Dockerfile [hunk 2]
```diff
@@ -18,28 +18,28 @@ FROM ${BASE_IMAGE}
...
-        openssh-server \
-        python3-dev \
-        rapidjson-dev \
-        sudo \
-        tmux \
-        unzip \
-        zstd \
-        zip \
-        zsh && \
+    autoconf \
```
Comment (sfc-gh-bprosnitz) [inline docker/Dockerfile:21]: Autoformat

File: docker/Dockerfile [hunk 3]
```diff
@@ -51,21 +51,37 @@ RUN mkdir -p /workspace/build/fastertransformer_backend/build
...
 RUN CUDAFLAGS="-include stdio.h" cmake \
-      -D CMAKE_EXPORT_COMPILE_COMMANDS=1 \
-      -D CMAKE_BUILD_TYPE=Release \
-      -D ENABLE_FP8=OFF \
-      -D CMAKE_INSTALL_PREFIX=/opt/tritonserver \
-      -D TRITON_COMMON_REPO_TAG="r${NVIDIA_TRITON_SERVER_VERSION}" \
-      -D TRITON_CORE_REPO_TAG="r${NVIDIA_TRITON_SERVER_VERSION}" \
-      -D TRITON_BACKEND_REPO_TAG="r${NVIDIA_TRITON_SERVER_VERSION}" \
-      .. && \
+    -D CMAKE_EXPORT_COMPILE_COMMANDS=1 \
```
Comment (sfc-gh-bprosnitz) [inline docker/Dockerfile:54]: Auto format

File: isolated-batcher/src/isolated_batcher.cc [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+/// about this pending batch. When the callback returns, this should reflect
+/// the latest batch information.
+/// \param should_include The pointer to be updated on whether the request
+/// should be included in the batch.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONBACKEND_ISPEC TRITONSERVER_Error*
+TRITONBACKEND_ModelBatchIncludeRequest(
+    TRITONBACKEND_Request* request, void* userp, bool* should_include)
+{
+  *should_include = false;  // Do not batch together multiple requests.
```
Comment (sfc-gh-bprosnitz) [inline isolated-batcher/src/isolated_batcher.cc]: Key line

**Review by sfc-gh-bprosnitz — COMMENTED** [review]
.

File: docker/Dockerfile (hunk 1 above)
Comment (sfc-gh-zhwang) [reply docker/Dockerfile:42]: weird, didn't see this error before

**Comment by sfc-gh-bprosnitz** [comment]
https://github.com/neevaco/neeva/pull/106482

File: corvo-batcher/src/corvo_batcher.cc [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+// - Limits the number of concurrent requests.
+
+#include <memory>
+#include <mutex>
+
+#include "triton/core/tritonbackend.h"
+
+namespace {
+
+constexpr uint64_t kNumConcurrentRequest = 16;
```
Comment (sfc-gh-ybsat) [inline corvo-batcher/src/corvo_batcher.cc:12]: this is ideally a value we set in the model config, as we will need to set it seperately for each model

**Comment by sfc-gh-ybsat** [comment]
This only sets a max queue size for the fastertransformer backend.
If we have a model ensemble, with some of the upstream models using a pytorch or python backend, then a request will be processed through those models first, using up resources along the way, then finding out it cant run any more once it hits the fastertransformer backend.

Note I am assuming this is the behavior, we will need to test it further. You can use my loadtest client here https://github.com/neevaco/neeva/pull/105895 which would give you the execution steps of a request in triton per model. If say for Llama2 we see that we get traces for the tokenizer model, but not the decoder model, then we validate this hypothesis.

If the above is true, we should look into how we can enforce queue size upstream, either in the beggining of the model ensemble, or at triton server level

File: corvo-batcher/src/corvo_batcher.cc [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+// Semaphore implements a basic semaphore in C++, similar to
+// std::counting_semaphore in C++20.
+class Semaphore {
```
Comment (sfc-gh-anzhang) [inline corvo-batcher/src/corvo_batcher.cc:16]: Have we considered other libraries w/ semaphore i.e. boost? This seems fine for now given that it's simple

File: corvo-batcher/src/corvo_batcher.cc [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    std::unique_lock lock(mu_);
+    n_++;
+  }
+
+ private:
+  std::mutex mu_;
+  uint64_t n_;  // protected by mu_
+};
+
+class CorvoBatcher {
```
Comment (sfc-gh-anzhang) [inline corvo-batcher/src/corvo_batcher.cc:41]: We can probably unit test this class, but up to you

**Comment by sfc-gh-zhwang** [comment]
> If the above is true, we should look into how we can enforce queue size upstream, either in the beggining of the model ensemble, or at triton server level

I think we can set this for every backend, not only fastertransformer. But you are right that, other models are super fast, so the queue size on other models shouldn't grow too big.

**Comment by sfc-gh-bprosnitz** [comment]
Closing because this doesn't work the way we thought

**Pull request closed without merging**